### PR TITLE
Use skia golden files in driver test

### PR DESCRIPTION
**Pull request by blasten**
## Description

Use Skia golden files in driver test, so we don't need to check in binaries into this repo. To make this happen, I removed the dependency on `dart:ui` from `flutter_test/lib/src/goldens.dart` and `flutter_test/lib/src/_goldens_io.dart` since  `dart:ui` isn't available in the vm.

I filed https://github.com/flutter/flutter/issues/49749 to track access to the Skia gold service from LUCI.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_test/lib/src/goldens.dart
Comment (jonahwilliams) [inline packages/flutter_test/lib/src/goldens.dart]: Though its useful for debugging, I don't think this is a good long-term change since we lose type safety on the APIs. I would recommend exploring how you can expose the same logic without changing the existing APIs

Comment (blasten) [reply packages/flutter_test/lib/src/goldens.dart]: maybe instead of having conditional import in this file, it's really up to the caller to either import web or io. That will remove this problem too. WDYT?

Comment (blasten) [reply packages/flutter_test/lib/src/goldens.dart]: done. PTAL

**Comment by Piinks** [comment]
Taking a look now, I've updated #49749 regarding authenticating Gold in the Luci environment, goldctl has a luci auth method that does not require a service account.

fyi @kjlubick :)

**Comment by Piinks** [comment]
This is going to be great to have tests running on device. Thank you! 🙂 These driver tests are executed in post-submit, right? 

I am curious about a few things here. First, I'd like to check which comparator is being used when these run. Right now there are couple of different comparators that may or may not be using Gold based on where/when the test is running. See https://github.com/flutter/flutter/blob/9375377fa99e8faae8a573e9fe92b8712d33e82d/packages/flutter_goldens/lib/flutter_goldens.dart#L31

Second, right now the skia gold/post-submit comparator will not return a negative test result, since we don't want to turn the tree red just for an intentional change. https://github.com/flutter/flutter/blob/9375377fa99e8faae8a573e9fe92b8712d33e82d/packages/flutter_goldens_client/lib/skia_client.dart#L257 (I think this can actually be changed now since we are checking in golden files in pre-submit with tryjobs.) 

If a change lands on master for these driver tests, what do we want to have happen? We won't know about it until it lands, because we aren't running pre-submit checks, so do we want the tree to turn red? Do we want to leave it green, but receive a comment on the PR that landed the change?

Before we decided to check in files in pre-submit, I had considered adding a benchmark for un-triaged results on master to bring them to our attention when they land. WDYT?

**Comment by blasten** [comment]
Thanks for the summary @Piinks!

My understanding is that we will start running presubmit tests on LUCI. (Some tests currently run on LUCI if someone adds the tag CQ+1 to a PR)

That said, I think it would make sense to follow the same procedure as any other golden test.

In the meanwhile, (for testing only) I can manually update the golden files. 

wdyt?

**Comment by Piinks** [comment]
That SGTM. :)  If we are getting ready to make the switch to luci for pre-submit, we'll just need to make a few changes to the cirrus-specific aspects of the goldenFileComparator. IIRC, right now for luci it just defers to the local comparator. I'm looking into some of that now, maybe we can touch base when I am back in-office next week? :)

**Comment by blasten** [comment]
Perfect. In the meanwhile, would you mind reviewing this PR, so once the changes land we can start testing the LUCI flow? I can manually update the golden files from a LUCI machine.

File: packages/flutter_test/lib/src/buffer_matcher.dart
Comment (Piinks) [inline packages/flutter_test/lib/src/buffer_matcher.dart]: You mentioned handling this manually until the client is ready for luci testing. It looks like in post-submit on luci this comparison is a no-op: https://github.com/flutter/flutter/blob/0be9127dfd934b8fe4bb363ea47d895d1e9830f9/packages/flutter_goldens/lib/flutter_goldens.dart#L489

**Comment by Piinks** [comment]
Also, is `goldctl` available on luci? Currently we download the binary for Cirrus testing on mac and windows, and it is in the docker file for linux tests.

**Comment by kjlubick** [comment]
Yes, goldctl is available from CIPD https://chrome-infra-packages.appspot.com/p/skia/tools/goldctl

File: packages/flutter_test/lib/src/buffer_matcher.dart
Comment (blasten) [reply packages/flutter_test/lib/src/buffer_matcher.dart]: Yep. I can definitely follow up with that based on the discussion on #49749.

Comment (Piinks) [reply packages/flutter_test/lib/src/buffer_matcher.dart]: Ok. Is this an already existing test? If so, idk if changing it to a no-op test in the interim is a good idea.

**Review by Piinks — COMMENTED** [review]
The larger re-factoring LGTM, but I don't know if we want to temporarily skip the actual test that this is being refactored for while we circle back to update the skia client/comparator.

File: dev/integration_tests/flutter_driver_screenshot_test/test_driver/main_test.dart
Comment (Piinks) [inline dev/integration_tests/flutter_driver_screenshot_test/test_driver/main_test.dart]: nit: Can we add a comment or TODO linking to the related issues about the no-op test?

**Review by Piinks — COMMENTED** [review]
Touched base with @blasten offline regarding the no-op tests. The pre-existing one will be restored and the new one will be no-op while we stand up the luci support to connect it to.

File: dev/integration_tests/flutter_driver_screenshot_test/test_driver/main_test.dart
Comment (blasten) [reply dev/integration_tests/flutter_driver_screenshot_test/test_driver/main_test.dart]: Done!

**Comment by blasten** [comment]
@Piinks PTAL

File: dev/integration_tests/flutter_driver_screenshot_test/README.md
Comment (Piinks) [inline dev/integration_tests/flutter_driver_screenshot_test/README.md]: Just a note to remember as we make the follow-up changes to this, we should see if adding device and environment parameters from these test environments will be beneficial for indexing device lab tests on Gold. 
e.g. https://github.com/flutter/flutter/blob/effc01905452a5d67a1e3bd9a028520592adffd2/packages/flutter_goldens_client/lib/skia_client.dart#L584
I may incorporate it into https://github.com/flutter/flutter/pull/49815

**Review by Piinks — APPROVED** [review]
LGTM! :)

**Pull request merged**
